sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs

```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
                     .ok_or(TransportError::InvalidOperation)?,
             );
         }
+
+        if let Some(cmsis_interface) = self.cmsis_interface {
+            // HyperDebug firmware supports binary protocol for retrieving list of events, use
+            // that for greatly improved performance.
+
+            let mut pkt = Vec::<u8>::new();
+            pkt.write_u8(Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO)?;
+            pkt.write_u8(Self::GPIO_MONITORING_READ)?;
+            pkt.write_u8(pin_names.len().try_into()?)?;
+            for pin_name in &pin_names {
+                pkt.write_u8(pin_name.len().try_into()?)?;
+                pkt.extend_from_slice(pin_name.as_bytes());
+            }
+            self.inner
+                .usb_device
+                .borrow()
+                .write_bulk(cmsis_interface.out_endpoint, &pkt)?;
+
+            let mut resp = RspGpioMonitoringHeader::new();
+            let mut bytecount = 0;
+            while bytecount < GPIO_MONITORIN_HEADER_SIZE {
+                let read_count = self.inner.usb_device.borrow().read_bulk(
+                    cmsis_interface.in_endpoint,
+                    &mut resp.as_bytes_mut()[bytecount..][..64],
+                )?;
+                ensure!(
+                    read_count > 0,
+                    TransportError::CommunicationError("Truncated GPIO response".to_string())
+                );
+                bytecount += read_count;
+            }
+            ensure!(
+                resp.encapsulation_header == Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO,
+                TransportError::CommunicationError(
+                    "Unrecognized CMSIS-DAP response to GPIO request".to_string()
+                )
+            );
+            match resp.status {
+                0 => (),
+                n => bail!(TransportError::CommunicationError(format!(
+                    "HyperDebug error: {}",
+                    n
+                ))),
+            }
```

Comment:
It just happens that I added handling for one specific error code, which deserves a descriptive error message, since it can be triggered by the operator, and not only by bugs in opentitantool/HyperDebug.